spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/ComposedTaskValidatorVisitor.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.core.dsl;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Visitor for a parsed composed task that verifies it is coherent. Rules being checked:
+ * <ul>
+ * <li>Any secondary sequences are labeled (otherwise they are unreachable).
+ * <li>Cannot label two things with the same string.
+ * <li>All target labels used on transitions must exist.
+ * <li>TODO much more!
+ * </ul>
+ *
+ * @author Andy Clement
+ */
+public class ComposedTaskValidatorVisitor extends ComposedTaskVisitor {
+
+	// Text of the AST being validated
+	private String composedTaskText;
+
+	private List<ComposedTaskValidationProblem> problems = new ArrayList<ComposedTaskValidationProblem>();
+
+	// At the end of the visit, verify any sequences that are never used
+	private List<LabelledComposedTaskNode> recordedSequences = new ArrayList<LabelledComposedTaskNode>();
+	
+	private Set<Transition> transitionsTargetingLabels = new HashSet<>();
+	
+	private Set<String> labelsDefined = new HashSet<>();
+
+	public List<ComposedTaskValidationProblem> getProblems() {
+		return problems;
+	}
+	
+	public boolean hasProblems() {
+		return problems.size()!=0;
+	}
+	
+	public void reset() {
+		this.problems.clear();
+	}
+	
+	public void startVisit(String composedTaskText) {
+		this.composedTaskText = composedTaskText;
+	}
+	
+	@Override
+	public boolean preVisitSequence(LabelledComposedTaskNode firstNode, int sequenceNumber) {
+		if (sequenceNumber > 0 && !firstNode.hasLabel()) {
+			pushProblem(firstNode.getStartPos(), ComposedTaskValidationProblem.SECONDARY_SEQUENCES_MUST_BE_NAMED);
+		}
+		recordedSequences.add(firstNode);
+		return true;
+	}
+
+	@Override
+	public void postVisitSequence(LabelledComposedTaskNode firstNode, int sequenceNumber) {
+	}
+
+	@Override
```

Comment:
Done. Actually they were left there because there is still so much more the validator should probably do and some of these would be the hooks. But I'm putting off that work for now, so I'll remove them at the moment.